### ISSUE TITLE
Pocket signin prompts for password even if I'm logged in to fxa

### ISSUE THREAD

**Issue by rfk** [mozilla/fxa-content-server#2391]
Steps to reproduce:

* Go to https://accounts.firefox.com and sign in
* Go to https://getpocket.com/ff_signin

Expected (and this actually worked in my default profile):

![screen shot 2015-05-08 at 09 25 27](https://cloud.githubusercontent.com/assets/34695/7527904/3a3ff5a0-f564-11e4-9c7a-374c6a566d6b.png)

Actual (this is what happened in a fresh profile):

![screen shot 2015-05-08 at 09 25 36](https://cloud.githubusercontent.com/assets/34695/7527906/409951c6-f564-11e4-8767-77f18b04ece8.png)

**Comment by vladikoff** [comment]
@rfk AFAIK it will always prompt for password **unless** you are signed in to sync in the browser.

**Comment by gavinsharp** [comment]
I am signed in to Sync in the browser, and I still got the password prompt.
(How would the page know I was signed in to Sync in the browser anyways?)

**Comment by zaach** [comment]
The content server keeps your credentials in localStorage after you sign in to Sync. If you've clicked "Use a different account" or cleared localStorage it will erase those credentials and you'll be prompted for a password again.

**Comment by rfk** [comment]
AFAICT I've got a sessionToken in my localStorage:
```
> localStorage.getItem("__fxa_storage.accounts")
"{"dc9f6b1846454688aebe5003d2766f8f":{"uid":"dc9f6b1846454688aebe5003d2766f8f","email":"ryan@rfk.id.au","sessionToken":"XXXYYYZZZ","accessToken":"AAABBBCCC","verified":true,"lastLogin":1431047428720}}"

> localStorage.getItem("__fxa_storage.currentAccountUid")
""dc9f6b1846454688aebe5003d2766f8f""
```

Should this be sufficient to trigger the passwordless-continue prompt?

**Comment by rfk** [comment]
Thanks to @zaach's pointers in IRC, I see what we also need a special `sessionTokenContext` value in the account data, and it currently must match sync in order to proceed without password prompt.

**Comment by zaach** [comment]
For reference: we decided on this distinction in #1621

**Comment by rfk** [comment]
Thanks @zaach and @vladikoff.  It's now clear that this is currently operating as expected, and I think we all understand the reasons why.

@gavinsharp mentions this as a potential blocker for pocket release, so let's burn a few brain cycles reconsidering #1621 in light of this new integration and see if we can come up with something.  /cc @ckarlof and @ryanfeeley.

Regardless of whether we decide on any changes to this flow, we should try to measure the relative frequency with which users have to enter their password.  I filed #2393 for this purpose.

**Comment by rfk** [comment]
Since the underlying problem here (from #1621) is approximately "logging out of a Firefox service doesn't log you out of FxA", perhaps we can consider making it do just that?  How gross would it be to get e.g. Loop signout to actually zap some fxa state and make it so the user must re-enter their password?  (I'm betting "pretty damn gross", like having them share cookie state on a common domain suffix, but...?)

I'll also note that we do have a "Sign Out" link on the front page of https://accounts.firefox.com, but it seems unlikely that users will know to go look for it there.

**Comment by rfk** [comment]
I think a sensible path forward is to try to land #1621 for train-37, and gather some metrics on how much this "signed in, but not through sync" state happens in practice while in nightly/beta.  That'll let us make an informed choice about changes we want to make, if any, and they can ship in train-38 and be (just) in time for pocket general release.

**Comment by ckarlof** [comment]
**The tl;dr is that if the user has logged into Sync previously, then she shouldn't need to re-enter her password to use Pocket.** [1] Since the vast majority of our FxA accounts are Sync users, this probably means a good experience for most users.

[1] Except if she's ever selected "Use different account" on the login screen. Without a full fledged account choose, we decided the simplest thing to do here for a coherent UX was to delete the cached credentials for the user in this case. 

The longer story:

This is a complex issue that we don't have all the answers to. Part of what makes it complex is:

1) We want to avoid re-asking the user for her password as much as possible.
2) We need a clear affordance to "sign out", which would stop password remembering. 

This complicated by the fact that FxA has relying services on the Web (FMD, Marketplace), attached to the Desktop browser (Sync, Hello), and now third party services (Pocket).

Here's the simple rule we settled on: **If logging in to firefoxServiceA enables passwordless login to subsequent services, then the sign out mechanism for firefoxServiceA should stop passwordless login.**

Otherwise, you get WTF experiences which are of the form "I logged out. Why does it appear that I'm still logged in?" See #1621 (i.e., Unable to fully sign out of relying services when credentials are cached). Not being able to log out is a frustrating and surprising experience. 

What this policy means in practice is that if you're signed to Sync, you don't need to enter your password again until you disconnect from Sync. I.e., "disconnect from Sync" is our clear logout mechanism, and it was chosen because the vast majority of our FxA users are Sync users. Unlike Google, we don't have any high traffic Web properties, so it doesn't make to put a logout mechanism there. Users would never discover it. 

So what about Hello? Lots of interesting questions here. We decided to decouple the logged in states of Hello and Sync, meaning the user can choose to be logged in to them independently. What about Pocket? Pocket is a third-party, and maintains its own logged in state. Its logged in and logged out state should be decoupled from the rest of our services. 

In general, we tried to keep it simple until we had more services connected to FxA, so we could learn more about what the problems were in practice. Perhaps it's time to revisit this. It could be that something like the Android account model would make sense here where you can tick and untick the services you want to connect with your FxA. I know that @ryanfeeley has thought about this before.

**Comment by rfk** [comment]
> The tl;dr is that if the user has logged into Sync previously, then she shouldn't need to re-enter her password to use Pocket.

And to be super clear: we believe this is currently working and the vast majority of existing FxA users will get this experience today, per @vladikoff's nice screencast: http://v14d.com/g/no-pass.gif

The metrics in  #2393 are to help verify that assertion.

**Comment by ckarlof** [comment]
I like triple knots on my shoelaces, too! Thanks, @rfk, @vladikoff, @zaach!

**Comment by vladikoff** [comment]
@rfk what action are we taking here for this train 38?

**Comment by rfk** [comment]
We need to compare the rate of `ask-password.shown.session-from-web` to `ask-password.skipped` events, to see what proportion of users are being denied as passwordless login in the current status quo.

**Comment by rfk** [comment]
@vladikoff has created a graph for this in datadog; thanks Vlad!

The initial comparison of `shown.session-from-web` to `skipped` is not as we expected it to be.  We're seeing around 4 or 5 `shown.session-from-web` events for every one `skipped` event - meaning that for every user who gets a passwordless login flow, there are 4 other users who *could* have got one, but didn't because their session was from something other than sync.

We need to get a bit more data into the system before drawing any conclusions here.  It could still be being skewed by devs and testing.  But I'm surprised by the initial results here.

**Comment by rfk** [comment]
After a few weeks for the data to settle, the graph still shows at least 50% of potential passwordless logins being denied due to this session-management policy.  But I'm wondering if there's a problem with the metrics.  Here are the first two logs I found in kibana containing the `session-from-web` event that indicates a lost opportunity for passwordless login:
```
events:  signin.ask-password.shown.account-unknown,screen.signin,error.signin.profile.100,signin.profile_image_not_shown,signin.password.visible,signin.submit,signin.success,settings.communication-prefs-link.visible.false,screen.settings,settings.profile_image_not_shown,settings.submit,settings.signout.submit,settings.signout.success,signin.ask-password.shown.account-unknown,screen.signin,error.signin.profile.100,signin.profile_image_not_shown,signin.submit,signin.success,settings.communication-prefs-link.visible.false,screen.settings,settings.profile_image_not_shown,signin.ask-password.shown.session-from-web,screen.signin,signin.profile_image_not_shown,settings.communication-prefs-link.visible.false,screen.settings,settings.profile_image_not_shown,signin.ask-password.shown.session-from-web,screen.signin,signin.profile_image_not_shown,signin.password.visible,signin.submit,signin.success,settings.communication-prefs-link.visible.false,screen.settings,settings.profile_image_not_shown,signin.ask-password.shown.session-from-web,screen.signin,signin.profile_image_not_shown
``

And
``
Events: oauth.signup.email-optin.visible.true,screen.oauth.signup,error.oauth.signup.auth.1023,error.oauth.signup.auth.1023,error.oauth.signup.auth.1023,oauth.signin.ask-password.shown.account-unknown,screen.oauth.signin,error.oauth.signin.profile.100,oauth.signin.profile_image_not_shown,oauth.signin.submit,screen.signin-permissions,oauth.signin.ask-password.shown.session-from-web,screen.oauth.signin,oauth.signin.profile_image_not_shown,oauth.signin.submit,screen.signin-permissions,oauth.signin.ask-password.shown.session-from-web,screen.oauth.signin,oauth.signin.profile_image_not_shown,oauth.signin.password.visible,oauth.signin.submit,screen.signin-permissions,oauth.signin.ask-password.shown.session-from-web,screen.oauth.signin,oauth.signin.profile_image_not_shown,oauth.signin.use-different-account,oauth.signin.ask-password.shown.account-unknown,oauth.signin.submit,screen.signin-permissions,signin-permissions.submit,signin-permissions.accept,signin-permissions.success
``

Each of these started out with prompting for password because of `account-unknown`, but went on to log several `session-from-web` events which would each count as part of our "we failed to offer a passwordless login experience" metric.

I guess these are being triggered by a re-render of the view logging spurious additional events?

**Comment by rfk** [comment]
@shane-tomlinson points out that there are multiple sign-in events in these flows, so maybe some sort of going-back-and-retrying behaviour on the part of the user here, rather than re-render.

**Comment by rfk** [comment]
Closing this in favour of the linked follow-up bugs, I don't think there's any more actions we can take in this bug
